File: old/session07LinkedList/LinkedList2022S.java
public class LinkedList2022S {
	private class Node { // LinkedList2022S.Node
		int val;
		Node next;// 32bits * 8 = 4* 10^9 * 8 = 32Gb
	}

	private Node head;
	private Node tail;
	public LinkedList2022S() {
		head = null;
	}

	public void addStart(int v) {
		Node temp = new Node();
		temp.val = v;
		temp.next = head; // put yourself in front of the first element
		head = temp;
	}

	public void addEnd(int v) {
		if (head == null) {
			head = new Node();
			head.val = v;
			head.next  = null;
			return;
		}
		Node p;
    for (p = head; p.next != null; p = p.next)
			;
		//p = pointing to last element
		Node temp = new Node();
		temp.val = v;
		temp.next = null;
		p.next = temp;
	}

	public void removeStart() {
    head = head.next;
	}

	public void removeEnd() {
    Node p, q;

		if (head == null) {
			return; // What?
		}
	 
		p = head;
		if (p.next == null) {
      head = null;
			return;
		}
		// invariant: at least 2 elements
		for (q = p.next; q.next != null; p = q, q = q.next)
			;
		p.next = null;
	}
	
	public String toString() {
		//NOTE: This is the bad way! Use StringBuilder
		String s = "";
		for (Node p = head; p != null; p = p.next)
			s = s + p.val + ",";
		return s;
	}

	public static void main(String[] args) {
		LinkedList2022S list = new LinkedList2022S();
		for (int i = 0; i < 10; i++)
			list.addStart(i);
		for (int i = 0; i < 10; i++)
			list.addEnd(i);
		list.removeStart();
		list.removeEnd();
		System.out.println(list);
	}
}

File: old/session07LinkedList/LinkedList2022S.cc
#include <iostream>
using namespace std;

class LinkedList {
private:
	struct Node {
	public:
    // by default, PRIVATE

		int val; // 4 bytes padding
		Node* next;
	};

	Node* head;
public:
	LinkedList() {
    head = nullptr;    
	}
	~LinkedList() {
		Node* temp;
		for (Node* p = head; p != nullptr;) {
			temp = p;
			p = p->next;
			delete temp;
		}
	}
	// copy constructor
	LinkedList(const LinkedList& orig) {

	}
	LinkedList& operator=(const LinkedList& orig) = delete;

	void addStart(int v) {
    Node* t = new Node();
		t->val = v;
		t->next = head;
		head = t;
	}
	
	void addEnd(int v) {
		if (head == nullptr) {
			head = new Node();
			head->val = v;
			head->next = nullptr;
			return;
		}
		
    Node* p;
		for (p = head; p->next != nullptr; p = p->next)
			;
		Node* temp = new Node();
		p->next = temp;
		temp->next = nullptr;
		temp->val = v;
	}
	void removeStart() {
		Node* temp = head;
		head = head->next;
		delete temp;
	}

	void removeEnd() {

	}
	friend ostream& operator <<(ostream& s, const LinkedList& list) {
		for (Node* p = list.head; p != nullptr; p = p->next)
			s << p->val << ' ';
		return s;
	}
};

void f() {
	LinkedList a;
	for (int i = 0; i < 10; i++)
		a.addStart(i);

}
int main() {
	LinkedList a; // create an empty LinkedList
	LinkedList b(); // b is a FUNCTION that returns LinkedList

	a.addEnd(3);
	for (int i = 0; i < 10; i++)
		a.addStart(i);
	for (int i = 0; i < 10; i++)
		a.addEnd(i);
	//	b.addStart(3);
	a.removeStart();
	a.removeEnd();
	cout << a << '\n';
}
